Client/Frontend/Browser/SearchLoader.swift

```diff
@@ -32,37 +32,64 @@ class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor<Site>, SearchViewController
 
     fileprivate lazy var topDomains: [String] = {
         let filePath = Bundle.main.path(forResource: "topdomains", ofType: "txt")
-        return try! String(contentsOfFile: filePath!).components(separatedBy: "\n")
+        do {
+            return try String(contentsOfFile: filePath!).components(separatedBy: "\n")
+        } catch {
+            Sentry.shared.send(message: "SearchLoader topDomains failed", tag: .general, severity: .error)
+            return []
+        }
     }()
 
+
+    private weak var currentDbQuery: Cancellable? = nil
+
     var query: String = "" {
         didSet {
+            guard let profile = self.profile as? BrowserProfile else {
+                assert(false)
+                return
+            }
+
             if query.isEmpty {
                 self.load(Cursor(status: .success, msg: "Empty query"))
                 return
             }
 
-            if let inProgress = inProgress {
-                inProgress.cancel()
-                self.inProgress = nil
+            if let deferred = currentDbQuery {
+                profile.db.cancel(deferred)
             }
 
             let deferred = self.profile.history.getSitesByFrecencyWithHistoryLimit(100, bookmarksLimit: 5, whereURLContains: query)
-            inProgress = deferred as? Cancellable
+            currentDbQuery = deferred as? Cancellable
 
-            deferred.uponQueue(DispatchQueue.main) { result in
-                self.inProgress = nil
+            deferred.upon() { result in
+                guard let deferred = deferred as? Cancellable, !deferred.cancelled else {
+                    return
+                }
 
                 // Failed cursors are excluded in .get().
-                if let cursor = result.successValue {
-                    // First, see if the query matches any URLs from the user's search history.
-                    self.load(cursor)
-                    for site in cursor {
-                        if let url = site?.url,
-                               let completion = self.completionForURL(url) {
+                guard let cursor = result.successValue else {
+                    return
+                }
+
+                // First, see if the query matches any URLs from the user's search history.
+                self.load(cursor)
+                for site in cursor {
+                    if deferred.cancelled {
+                        return
+                    }
+
+                    if let url = site?.url,let completion = self.completionForURL(url) {
+                        DispatchQueue.main.async {
                             self.urlBar.setAutocompleteSuggestion(completion)
-                            return
                         }
+                        return
+                    }
+                }
+
+                DispatchQueue.main.async {
+                    if deferred.cancelled {
```

Comment:
How 'bout we do most of the work that's inside this block, apart from the final `urlBar` touch, off the main thread?